Simplify property and add type hints

rwb27 opened the pull request:
This PR significantly reorganises how properties work. The functionality is split into separate classes with a common base, and there are fewer, better-defined ways for a property to work.

* `DataProperty` is a variable-like property. It may not have a getter/setter, and functions just like a variable. You may specify whether it can be written to from client code, that's currently its only option.
* `FunctionalProperty` uses getter/setter. It works pretty much exactly like Python's `property`.
* Both these classes are used via a common function, `lt.property`. See example below.
* Documentation-related stuff (remembering the name, generating a title/description) is moved to a separate base class. In due course, this could also be used for actions.
* `lt.setting` functions just like `lt.property`, but the values are remembered in the settings file for the `Thing`. The implementation of saving/loading settings has not changed.

The two ways for defining properties now look like:

```python
import labthings_fastapi as lt

class Counter(lt.Thing):
    "A counter that knows what's remaining."

    count: int = lt.property(default=0, readonly=True)
    "The number of times we've increnented the counter."

    target: int = lt.property(default=10)
    "The number of times to increment before we stop."

    @lt.property
    def remaining(self) -> int:
        "The number of steps remaining."
        return self.remaining - self.count

    @remaining.setter
    def set_remaining(self, value: int) -> None:
        self.target = self.count + value
```

I am happy with this syntax. It's not exactly what I wanted, but I think it is the best compromise between elegance and practicality. The two quirks I would like to draw attention to are:

* Both `default` and `default_factory` are keyword-only arguments. This enables `mypy` to distinguish between use as a decorator and use as a field, so that both cases are typed correctly.

  An earlier commit allowed default to be the (first and only) positional argument: it worked fine, but it confused `mypy`. We care about type checking in code that defines `Thing`s, so it's worth adopting the form that `mypy` can understand.
* When using `@lt.property` or `@lt.setting` as a decorator, `mypy` will raise an error if the same name is used for the getter and setter functions, because doing so means the descriptor is overwritten (with itself). The descriptor works in both cases.

There are quite a few things remaining to do before this is ready to be merged. These include:

- [x] Implement `BaseProperty.model` to ensure non-pydantic types can be wrapped in a `RootModel` when needed.
- [x] Allow settings with getters and setters
- [x] ~Add a way to attach an `on_set` function to a `DataProperty`. This would allow validation and synchronisation of `DataProperty` instances, which is required because settings are now `DataProperty` instances.~ This can be postponed, because getters/setters are back in.
- [x] Get all tests passing
- [x] Get mypy passing
- [x] Write tests for typing, ensuring mypy raises the right errors if properties are used incorrectly, and that it can infer the right types.
- [x] Support docstrings-after-attributes. Currently, data properties have no docstring.
- [x] Consider the module names carefully for `thing_property` and `base_descriptor`
- [x] Test `openflexure-microscope-server` with this branch.
- [x] Add tests to explore whether there's anything that might happen as a result of setters being named differently to getters.
- [x] Revisit inheriting from `property` just to make sure that doesn't solve one of our problems.
- [x] ~Add a documentation page on settings.~ moved to #171.

julianstirling commented:
> I've not yet fully worked through settings - that's a todo I should make more explicit. I think, though, that I am very keen to enforce variable-like behaviour for settings - and I think that means settings should not be implemented with getters and setters. 

I have found so far most settings I have needed to use getters and setters in the microscope code base. If getters and setters are not allowed then we need to make sure that the descriptors are poweful enogh for all use cases. For example we have a lot of things where the information is held in other classes so we need to be able to iterate through them on get and set. Without getters and setters we would need a complex signalling mechanism that would go wrong.

rwb27 commented:
> > I've not yet fully worked through settings - that's a todo I should make more explicit. I think, though, that I am very keen to enforce variable-like behaviour for settings - and I think that means settings should not be implemented with getters and setters.
> 
> I have found so far most settings I have needed to use getters and setters in the microscope code base. If getters and setters are not allowed then we need to make sure that the descriptors are poweful enogh for all use cases. For example we have a lot of things where the information is held in other classes so we need to be able to iterate through them on get and set. Without getters and setters we would need a complex signalling mechanism that would go wrong.

I should look through those and figure out if there's a way it would work nicely, or if my assumptions are just wrong. I'm aware of the ones in the camera - I will have a look in the microscope codebase for others, but might miss stuff if it's in branches. Pointing me at examples of where you've used getters/setters for settings would be helpful.

julianstirling commented:
> > > I've not yet fully worked through settings - that's a todo I should make more explicit. I think, though, that I am very keen to enforce variable-like behaviour for settings - and I think that means settings should not be implemented with getters and setters.
> > 
> > 
> > I have found so far most settings I have needed to use getters and setters in the microscope code base. If getters and setters are not allowed then we need to make sure that the descriptors are poweful enogh for all use cases. For example we have a lot of things where the information is held in other classes so we need to be able to iterate through them on get and set. Without getters and setters we would need a complex signalling mechanism that would go wrong.
> 
> I should look through those and figure out if there's a way it would work nicely, or if my assumptions are just wrong. I'm aware of the ones in the camera - I will have a look in the microscope codebase for others, but might miss stuff if it's in branches. Pointing me at examples of where you've used getters/setters for settings would be helpful.

I would just do a search of `@lt.thing_setting` on the v3 branch. But I would strongly recommend not stopping the use of setters and getters for settings for three reasons:

* It breaks parity between properties and settings. The nice thing about setting is you change a property to a setting and it saves. If we property have ways of being created that settings don't then it makes it much harder to make something persist.
* Settings tend to be more complex. There is often a complex functional data that needs to persist, often that is less important than being read over HTTP. Setting provides a very uniform way of saving state information. If settings have to be dumb then we will find ourselves in situations where we need to create other ways of saving state information.
* It gives us fewer ways to work around the fact that the permutation space of LabThings has not been explored in anger. I have found a number of times that I try to create something and due to how the date is captured and sent it causes a recursion error, or if the data type is vaguely know and so the a clear base model is not know, then it doesn't seriealise and deserialise correctly. Currently each time this happens we can create an issue in LabThings to improve it, but we have the options of switching to a getter and a setter to carry on with our development. If this is lost then we loose will not be able to proceed until there is a LabThings-FastAPI release

julianstirling commented:
Here is an example that I see becoming more and more common as we try to make the algorithms that Things use for certain tasks configurable:

<img width="849" height="633" alt="image" src="https://github.com/user-attachments/assets/0715f34e-a6cb-4743-835b-511b182347ae" />

BackgroundDetection is now done with one of many BackgroundDetector objects, each object is a different instance of a different class. These each have a defined their own data structure (via a BaseModel)  for their settings, and for the information they save about the background. This should be very flexible for adding new algorithms without changing any Thing code.

On init we create every dackground detector, but their settings and data are persistent, so we load the settings from disk and find each instance in turn and pass them their settings and data. This is something we will also need to do for path planners that are needed to allow us to switch how we plan the path for a scan, without creating a new Thing for each type of scan.

We need this working very soon as we need to scan over a pre-defined grid for the urine microscopy.

Just like with the namespace changes and the last changes to ThingSettings, before we merged them into the main branch on LabThings we created an example branch on the server to check how it performs in a more complex situation. As we add more and more unit and integration tests to the microscope this will be our best resource for checking that any large changes can support the behaviour we need.

rwb27 commented:
> I would just do a search of `@lt.thing_setting` on the v3 branch.

Of course - that is something I'll do, along with testing this branch together with a branch of the microscope server. Thanks for highlighting the background detect example though, that is a really useful one.

I am persuaded that trying to eliminate functional settings in this PR will cause a jam, so I'll add some multiple inheritance and work around it.

I will move discussion of whether functional settings are a good idea to an issue thread. I think it really depends on how we see the relationship between individual settings and the settings file. This changed significantly in #110 and I think we need to revisit some assumptions to make it clear how settings ought to work. Once that's well-defined, we can make sure the implementation matches the intention.

julianstirling commented:
That's a shame! Perhaps a conventions such as a leading underscore, with a docstring highlighting it in examples.

```python

import labthings_fastapi as lt

class MyThing(Thing):

    @lt.property
    def the_answer(self) -> int:
        return 42

    @the_answer.setter
    def _the_answer(self, val: int) -> None:
        """Note the leading underscore! Blah blah explanation"""
        if val != 42:
            raise ValueError("Don't be ridiculous.")
```

It may be possible to write a MyPy plugin to get around this for `thing_properties`

rwb27 commented on docs/source/documentation.rst:
```diff
@@ -0,0 +1,16 @@
+
+Each :ref:`wot_thing` is documented by a Thing Description, which is a JSON document describing all of the ways to interact with that Thing. The WoT_ standard defines the `Thing Description`_ and includes a JSON Schema against which it may be validated.
+
+Thing Description documents are higher-level than OpenAPI_ and focus on the capabilities of the Thing, rather than the HTTP endpoints. In general you would expect more HTTP endpoints than there are interaction affordances, so in principle client code based on a Thing Description should be more meaningful. However, OpenAPI is a much more widely adopted standard and so both forms of documentation a ...
```
I think we need an efficient way to describe "properties and actions, and events as and when they are implemented". I have added a reference target, so I can write ``:ref:`wot_affordances``` and it links to the "Interaction Affordances" section of the WoT core concepts page, which I hoped gets the point across, without having to write a whole sentence every time we refer to it (which is often). I am very open to better ways, but I would be keen not to have lots of words that will end up slightly inconsistent and out of date (e.g. when we add events).

I see this particular text does not make interaction affordance a hyperlink, so I'll do that - I hope that's helpful. I can also explain why each action/property has >1 endpoint.

rwb27 commented on docs/source/tutorial/properties.rst:
```diff
@@ -0,0 +1,111 @@
+.. _tutorial_properties:
```
My intention is to add another page on settings. I think that makes them easier to find/link to.

julianstirling replied:
This is now #171

rwb27 commented on docs/source/tutorial/properties.rst:
```diff
@@ -0,0 +1,111 @@
+
+        # Make the property read-only for clients
+        twice_my_property.readonly = True
+
```
I think that's mostly duplicating the sentence immediately before the code block, but I suppose that sentence doesn't explicitly say it's related to `twice_my_property` so I'll add that.

bprobert97 commented on tests/test_base_descriptor.py:
```diff
@@ -0,0 +1,76 @@
+
+
+def test_docstrings_are_cached():
+    """Check that the docstrings aren't being regenerated every time."""
+    docs1 = get_class_attribute_docstrings(Example)
+    docs2 = get_class_attribute_docstrings(Example)
+    assert docs1 is docs2
```
Not sure how this checks they aren't being regenerated when Example is statically defined at the top of the file. Is get_class_attribute_docstrings querying over http?

rwb27 replied:
`get_class_attribute_docstrings` returns a dictionary. If it was being regenerated, then `docs1 == docs2` would be true, but `docs1 is docs2` would not be true, because it's a new dictionary (containing the same values).

In order to get the docstrings, I need to manually get the source of the class and parse it to an AST, then extract them into a dictionary. Unfortunately, it's not automatic, like `__doc__` for a class or method.

Does that answer the question? I've added a docstring and a comment in c6b24e72925eab419b9fcee6717f5866e1e35d48.

bprobert97 commented on tests/test_base_descriptor.py:
```diff
@@ -0,0 +1,76 @@
+    assert p.description is None
+
+
+def test_basedescriptor_get():
```
Add docstring

bprobert97 commented on tests/test_settings.py:
```diff
@@ -96,6 +121,43 @@ def test_settings_save(thing, server):
+    setting_file = _get_setting_file(server, "/thing")
+    server.add_thing(thing, "/thing")
+    # No setting file created when first added
+    assert not os.path.isfile(setting_file)
+    with TestClient(server.app) as client:
+        r = client.put("/thing/boolsetting", json=True)
+        assert r.status_code == 201
+        r = client.get("/thing/boolsetting")
+        assert r.json() is True
+        assert os.path.isfile(setting_file)
+        with open(setting_file, "r", encoding="utf-8") as file_obj:
```
add some comments here to describe what is happening in the test (e.g. what is a 201 response?)

rwb27 replied:
Have done (and for the nearly-identical function above it, which I didn't change) in 531e87e2e8bdd4fdcf8c41722997f46bc8b55e9f.

rwb27 commented on typing_tests/README.md:
````diff
@@ -0,0 +1,11 @@
+The codebase is type-checked with `mypy src/` and tested with `pytest`, however neither of these explicitly check that `mypy` can infer the correct types for `Thing` attributes like properties and actions. The Python files in this folder are intended to be checked using:
+
+```terminal
+mypy --warn-unused-ignores .\typing_tests\
````
Thanks - done in efc330eb033cac7b2010d7f0a2b860a0437a6a79

rwb27 commented on tests/test_properties.py:
```diff
@@ -45,22 +48,35 @@ def toggle_boolprop_from_thread(self):
 
 
-def test_instantiation_with_model():
+def test_instantiation_with_model() -> None:
```
Thanks - I'm not quite sure why I felt the need to add a return hint to this one test!

rwb27 replied:
Removed in c92500f0b568e5025a1cf53c4473673412007e5c

rwb27 commented on tests/test_properties.py:
```diff
@@ -45,22 +48,35 @@ def toggle_boolprop_from_thread(self):
 
 
-def test_instantiation_with_model():
+def test_instantiation_with_model() -> None:
```
I've gone through this whole test module and added lots of comments and docstrings, in c92500f0b568e5025a1cf53c4473673412007e5c.

I also added a test, added a bunch more assertions, and fixed a race condition that should close #139.

rwb27 commented on .github/workflows/test.yml:
```diff
@@ -75,6 +75,9 @@ jobs:
         run: mypy src
+      
+      - name: Type tests with MyPy
+        run: mypy --warn-unused-ignores ./typing_tests/
```
This was fixed in efc330eb033cac7b2010d7f0a2b860a0437a6a79, thanks.

bprobert97 commented on src/labthings_fastapi/thing_property.py:
```diff
@@ -0,0 +1,988 @@
+) -> FunctionalSetting[Value] | Value:
+    r"""Define a Setting on a `.Thing`\ .
+
+    A setting is a property that is saved to disk
```
```suggestion
    A setting is a property that is saved to disk.
```

bprobert97 commented on src/labthings_fastapi/thing_property.py:
```diff
@@ -0,0 +1,988 @@
+
+    A setting can be accessed via the HTTP API and is persistent between sessions.
+
+    A `.ThingSetting` is a `.DataProperty` with extra functionality for triggering
```
```suggestion
    A `.DataSetting` is a `.DataProperty` with extra functionality for triggering
```

bprobert97 commented on src/labthings_fastapi/thing_property.py:
```diff
@@ -0,0 +1,988 @@
+
+    A setting can be accessed via the HTTP API and is persistent between sessions.
+
+    A `.ThingSetting` is a `.FunctionalProperty` with extra functionality for
```
```suggestion
    A `.FunctionalSetting` is a `.FunctionalProperty` with extra functionality for
```

julianstirling commented on src/labthings_fastapi/exceptions.py:
```diff
@@ -1,16 +1,54 @@
+class DocstringToMessage:
+    """A mixin to put Exception docstrings in as their default message."""
+
+    append_to_message: bool = True
+
+    def __init__(self, message: str | None = None):
+        """Initialise an error with a message or its docstring.
+
+        :param message: the optional message.
+        """
+        # We ignore call-arg within this function because mypy can't know
+        # that this is a mixin, and super() will be an exception (which does
+        # accept a string argument to `__init__`).
+        doc = inspect.cleandoc(self.__doc__) if self.__doc__ else None
+        if message:
+            if doc and self.append_to_message:
+                super().__init__(message + "\n\n" + doc)  # type: ignore[call-arg]
+            else:
+                super().__init__(message)  # type: ignore[call-arg]
+        elif doc:
+            super().__init__(doc)  # type: ignore[call-arg]
+        else:
+            super().__init__()
```
I think having the docstring in at the top of the error message pushes the actual error message down. Normally if a traceback is weird (which can happen, especially with errors during handline) I use the error message to try to find the code. This would mean that the error message wasn't as written, which just seems confusing.

I think we can debate the pros and cons of this sort of mixin, but I think it would be best done in a seperate MR

rwb27 replied:
I was enthusiastic about it at first, but I am no longer totally convinced it's helpful. Pasting the docstring in saves having to look it up, on the other hand now we have working API docs that's less of an issue, and I agree making the message too long can be very confusing.

I'm happy to take this out for now.

rwb27 replied:
deleted in 55dcb005e97b489c2097012ada8e6530aa7cba24.

julianstirling commented on tests/test_settings.py:
```diff
@@ -11,21 +11,41 @@
-        self._float = value
+        self._floatsetting = value
+
+    @lt.setting
+    def readonlysetting(self) -> str:
+        """A read-only setting."""
+        return "This is read-only"
+
+    @readonlysetting.setter
+    def readonlysetting(self, value: str):
+        pass
```
If we are testing readonly but with a setter it would be good to make it functional so that we can check it with more in-depth tests. 
"This is read-only"
```suggestion

    .self._readonlysetting: str = "This is read-only"

    @lt.setting
    def readonlysetting(self) -> str:
        """A read-only setting."""
        return self._readonlysetting

    @readonlysetting.setter
    def readonlysetting(self, value: str):
        self._readonlysetting = value
```

rwb27 replied:
Happy to make it functional. Is `.self.` a language feature I've missed? A quick google (and typing it into my python interpreter) doesn't leave me any the wiser...

julianstirling commented on src/labthings_fastapi/thing_property.py:
```diff
@@ -0,0 +1,988 @@
+def property(  # noqa: D103
+    getter: Callable[[Any], Value],
+) -> FunctionalProperty[Value]: ...
+@overload  # use as `field: int = property(0)``
```
Is this true? Isn't default needed?
```suggestion
@overload  # use as `field: int = property(default=0)``
```

I am not overly familiar with overload. Also what happens if default is not specified, does it try to set the getter to 0 and then error?

I very much agree this code could do with spacing. I have had to check it out locally to read it.

rwb27 replied:
Yes, if it's called with a single positional argument that is not callable, it will fail with an error. I'll add a test for that.

julianstirling commented on src/labthings_fastapi/thing_property.py:
```diff
@@ -0,0 +1,988 @@
+                status_code=201,
+                response_description="Property set successfully",
+                summary=f"Set {self.title}",
+                description=f"## {self.title}\n\n{self.description or ''}",
```
Is there a WoT related reason that the title goes at the top of the description? I was thinking that it might be good to use the description as a tooltip in the UI? If not we may want to have some syntax for our descriptions such as:

```
:tooltip: This shows up in the UI
```

rwb27 replied:
This `description` is for OpenAPI, not the Thing Description. I think I did that so that the OpenAPI documentation would look nice - without it, there's no title shown.

The `description` field in the Thing Description (see the `property_affordance` method) does not have anything prepended, and that's what you'd get if you queried the Thing Description that's already in the web app store.

Having tooltips in there somewhere would certainly be nice. One to think about in an issue, the PR is big already.

The Thing Description has arbitrary metadata fields we're not currently using, so (if the plain `description` is too verbose, which it sometimes will be) adding a tooltip in there would definitely be possible. Extracting the tooltip from the docstring feels very sensible.

rwb27 replied:
I guess `readonly` shows we can have extra attributes with both styles of property already, adding a `tooltip` would be absolutely doable when the time comes.

rwb27 commented on typing_tests/thing_definitions.py:
```diff
@@ -0,0 +1,206 @@
+    intprop4: int = lt.property()  # type: ignore[call-overload]
+    "This property should cause mypy to throw an error, as it has no default."
+
+    listprop: list[int] = lt.property(default_factory=list)
```
I'm testing it doesn't raise an error - `list` is not of type `list[int]` but should be assignable.

rwb27 commented on src/labthings_fastapi/thing_property.py:
```diff
@@ -0,0 +1,988 @@
+
+
+# See comment at the top of the file regarding ignored linter rules.
+@overload  # use as a decorator  @property
+def property(  # noqa: D103
+    getter: Callable[[Any], Value],
+) -> FunctionalProperty[Value]: ...
+@overload  # use as `field: int = property(0)``
+def property(*, default: Value, readonly: bool = False) -> Value: ...  # noqa: D103
+@overload  # use as `field: int = property(default_factory=lambda: 0)` # noqa: E302
+def property(  # noqa: D103
+    *, default_factory: Callable[[], Value], readonly: bool = False
+) -> Value: ...
+def property(
```
`ruff` and `flake8` disagree on the right spacing here, I tend to assume ruff will add or remove newlines as it sees fit, but maybe it doesn't have a fixed answer here. I have often seen overloads together as a block which is why I wrote it like that, but I'm happy to add some newlines. Is the usual convention of 2 newlines between top-level definitions applicable here? That would seem a lot of whitespace for something that's all part of one function - but I will do whatever's conventional.

julianstirling replied:
Ruff allows no lines or 2 lines but not one. I had to check out the file and space it out to be able to read what was going on, so I would push for 2 lines not none

rwb27 replied:
fair enough, will do.

rwb27 replied:
Done in 8b18d754bc3e5162d6849e0e7894cacb85d9f370.

rwb27 commented on typing_tests/thing_definitions.py:
```diff
@@ -0,0 +1,206 @@
+    intprop: int = lt.property(default=0)
+    optionalintprop: int | None = lt.property(default=None)
+    optionalintprop2: int | None = lt.property(default=0)
+    optionalintprop3: int | None = lt.property(default_factory=optional_int_factory)
```
Sure :)

julianstirling commented on src/labthings_fastapi/thing_property.py:
```diff
@@ -0,0 +1,988 @@
+    def fdel(self) -> None:  # noqa: DOC201
+        """The deleter function.
+
+        This function always returns ``None`` as deleters are not yet supported.
```
I think getting rid of it is best, I don't think deleting a ThingProperty make sense.

rwb27 replied:
Removed in 43e436f920a8e08cb4f6bde5252fccf603a1d447.

rwb27 commented on typing_tests/thing_definitions.py:
```diff
@@ -0,0 +1,206 @@
+    intprop5 = lt.DataProperty[int]()  # type: ignore[call-overload]
+    """This property should cause mypy to throw an error, as it has no default."""
+
+    optionalintprop1 = lt.DataProperty[int | None](default=None)
+    """A DataProperty that should not cause mypy errors."""
```
This test only checks whether types match or they don't. The actual type doesn't make much difference to that. I could add tests for `int`, `str`, `float`, etc. but it wouldn't be exploring any different behaviour, code paths, or overloads.

Do you think there is merit in adding other types even so? My worry was that by adding more types, I'd give the impression we needed to enumerate all the possible types and thus the test would be longer but look less complete.

It might be worth testing more types in the `pytest` test, which is actually executed - but this file is only looked at by `mypy`.

rwb27 commented on src/labthings_fastapi/thing_property.py:
```diff
@@ -0,0 +1,988 @@
+    *, default_factory: Callable[[], Value], readonly: bool = False
+) -> Value: ...
+def property(
+    getter: ValueGetter | EllipsisType = ...,
+    *,
+    default: Value | EllipsisType = ...,
+    default_factory: ValueFactory | None = None,
+    readonly: bool = False,
```
I've added a note in the "typing notes" section of the docstring.

rwb27 commented:
> Thanks for this @rwb27.
> 
> I have gone through this to a degree. I haven't gone through the code in `thing_property` in super fine detail. It is quite a lot lower level python than I am comfortable with so it will take some time.
> 
> I think spacing out the overloads as @bprobert97 suggested would make it more readable, and I have also asked some questions that are largely for my own understanding.
> 
> I have pulled it and run the tests locally and looked through the example code in the tests. I think at this stage I can agree that this syntax is nicer and we will want to use it. I am particularly happy to see that we have a `default_factory` as this is needed in quite a few places in the OpenFlexure code.

Yes, `default_factory` is long overdue.

> I am happy with the structure that we can proceed to building a test implementation in a branch of OpenFlexure to check for issues when it is used in a more complex setting?

Great, thanks. I have the intention that I'll use the new syntax in an OFM server branch, but I will be sure to ping the software chat when I start, because it would be really sad if we ended up doing it in parallel...

bprobert97 commented on tests/test_property.py:
```diff
@@ -0,0 +1,218 @@
+
+def test_default_factory_from_arguments():
+    """Check the function that implements default/default_factory
+
```
```suggestion
```

rwb27 replied:
Fixed in f3963dffb02d8af27444eac949febb0ff9a30c17.

rwb27 commented on typing_tests/thing_definitions.py:
```diff
@@ -0,0 +1,206 @@
+assert_type(test_defaults_match.intprop3, int)
+assert_type(test_defaults_match.optionalintprop, int | None)
+assert_type(test_defaults_match.optionalintprop2, int | None)
+assert_type(test_defaults_match.optionalintprop3, int | None)
```
Done in df8c5a9dfea5f7adb593c5eb9073a093d6cc46cd.

rwb27 commented on typing_tests/thing_definitions.py:
```diff
@@ -0,0 +1,206 @@
+
+
+class TestExplicitDescriptor(lt.Thing):
+    """A Thing that checks our explicit descriptor type hints are working."""
```
Done in df8c5a9dfea5f7adb593c5eb9073a093d6cc46cd, I even explained why.

rwb27 commented on typing_tests/thing_definitions.py:
```diff
@@ -0,0 +1,206 @@
+
+assert_type(test_explicit_descriptor.optionalintprop1, int | None)
+assert_type(test_explicit_descriptor.optionalintprop2, int | None)
+assert_type(test_explicit_descriptor.optionalintprop3, int | None)
```
Done in df8c5a9dfea5f7adb593c5eb9073a093d6cc46cd.

rwb27 commented on typing_tests/thing_definitions.py:
```diff
@@ -0,0 +1,206 @@
+assert_type(TestExplicitDescriptor.optionalintprop3, lt.DataProperty[int | None])
+
+Val = typing.TypeVar("Val")
+Thing = typing.TypeVar("Thing", bound=lt.Thing)
```
Done in b2f597eb7b705a9142fa3b7aebad18b68116b351.

julianstirling commented:
I have tested MR !348 of the OpenFlexure Microscope which uses the new syntax from this branch (and with this branch checked out). Everything works as expected.

The code looks much _much_ cleaner with the new syntax which is great.

@rwb27; this is still in draft. What do we need to get it in?

rwb27 commented:
OK, I realised that re-testing whether having `FunctionalProperty` inherit from `builtins.property` was actually really easy, thanks to the `typing_tests` I've introduced. I have added `builtins.property` as a base class of `FunctionalProperty` and I still get an error that the property is being redefined if the getter and setter have the same name.

It's possible some tools will work better if `FunctionalProperty` inherits from `builtins.property`, but `mypy` does not seem to be such a tool. I think, therefore, I can tick that box but not make any code changes.

Inheriting from `builtins.property` causes a bunch of problems (for example, it would reintroduce a deleter function which we don't want, and require several extra `type: ignore` statements). It might be possible to get around these, but unless there's a strong reason to do so I don't think it's worthwhile. Technically, `FunctionalProperty` is not compatible with `builtins.property` as a subclass, because it's more strictly typed in terms of the value (`builtins.property` uses a lot of `Any` and only works because `mypy` treats it as a special case).
